pkg/mirror-producer/producer_test.go

```diff
@@ -0,0 +1,888 @@
+
+			conf := fmt.Sprintf(`{
+				"cniVersion": "%s",
+				"name": "mynet",
+				"type": "ovs-cni-mirror-producer",
+				"bridge": "%s",
+				"mirrors": %s
+			}`, version, bridgeName, mirrorsJSONStr)
+
+			It("should successfully complete ADD, CHECK and DEL commands", func() {
+				targetNs := newNS()
+				defer func() {
+					closeNS(targetNs)
+				}()
+
+				By("create interfaces/ports using ovs-cni plugin")
+				prevResult1 := createInterfaces(IFNAME1, targetNs)
+				prevResult2 := createInterfaces(IFNAME2, targetNs)
+
+				By("run ovs-cni-mirror-producer passing prevResult")
+				confMirror1, result1 := testAdd(conf, mirrors, prevResult1, IFNAME1, targetNs)
+				confMirror2, result2 := testAdd(conf, mirrors, prevResult2, IFNAME2, targetNs)
+				testCheck(confMirror1, result1, IFNAME1, targetNs)
+				testCheck(confMirror2, result2, IFNAME2, targetNs)
+
+				// mirror must have both ports (hostIfaces obtained from 'prevResult1' and 'prevResult2')
+				By("verify that mirrors have all ports")
+				checkPortsInMirrors(mirrors, prevResult1, prevResult2)
+
+				By("run ovs-cni-mirror-producer to delete mirrors")
+				testDel(confMirror1, mirrors, result1, IFNAME1, targetNs)
+				testDel(confMirror2, mirrors, result2, IFNAME2, targetNs)
+			})
+		})
+	})
+
+	Context("adding a mirror with both producer and consumer configuration", func() {
+		Context("('output_port', 'select_src_port' and 'select_dst_port' defined with valid portUUIDs)", func() {
+			mirrors := []types.Mirror{
+				{
+					Name:    "mirror-prod",
+					Ingress: true,
+					Egress:  true,
+				},
+			}
+			mirrorsJSONStr, err := toJSONString(mirrors)
+			Expect(err).NotTo(HaveOccurred())
+
+			conf := fmt.Sprintf(`{
+				"cniVersion": "%s",
+				"name": "mynet",
+				"type": "ovs-cni-mirror-producer",
+				"bridge": "%s",
+				"mirrors": %s
+			}`, version, bridgeName, mirrorsJSONStr)
+
+			It("shouldn't be removed after calling cmdDel by this plugin, because it contains 'output_port' configured by a consumer", func() {
+				// This is very important:
+				// cmdDel of both mirror-producer and mirror-consumer plugins is able to cleanup a mirror
+				// without a useful configuration (all traffic outputs and inputs are undefined).
+				// However, they can remove a mirror only if both
+				// 'output_port', 'select_src_port' and 'select_dst_port' are empty.
+				targetNs := newNS()
+				defer func() {
+					closeNS(targetNs)
+				}()
+
+				By("create interfaces using ovs-cni plugin")
+				prevResult := createInterfaces(IFNAME1, targetNs)
+
+				By("run ovs-cni-mirror-producer passing prevResult")
+				confMirror, result := testAdd(conf, mirrors, prevResult, IFNAME1, targetNs)
+				testCheck(confMirror, result, IFNAME1, targetNs)
+
+				By("create a consumer interface and add its port via 'ovs-vsctl' to fill mirror 'output_port'")
+				r2 := createInterfaces(IFNAME2, targetNs)
+				portUUID := getPortUUIDFromResult(r2)
+				addOutputPortToMirror(portUUID, mirrors[0].Name)
```

Comment:
Do we need this helper here? Could this be tested by adding several mirror users, removing one, confirming that the mirror still exists, and continuing until there are no users and only then confirm the mirror is gone?

Reply:
We added this to create a more realistic test where a producer have also output_port (basically 99,99% of the times in real scenarios). cmdDel logic of mirror-producer must be able to cleanup a mirror only if also output-port is empty.
The same for mirror-consumer, because it can remove a mirror only if also select_* are empty.

Are you talking about *consumers* with "mirror users"?
If yes, the answer is no, because a mirror can have multiple producers but only a single consumer, in fact ovsdb spec states that `output_port` is an optional weak reference to Port, and not "set of weak reference...".
Instead, if you are talking about *producers*, that scenario is covered by "adding multiple ports to a single mirror", because there is a single mirror with 2 portUUIDS in select_*. However, we could improve this test checking mirror ports between the 2 cmdDel and at the end.

Reply:
Ah, thanks, I think I understand now